scvi/model/_multivi.py

```diff
@@ -0,0 +1,712 @@
+from scvi._utils import _doc_params
+from scvi.dataloaders import DataSplitter
+from scvi.model._utils import (
+    _get_batch_code_from_category,
+    _get_var_names_from_setup_anndata,
+    scatac_raw_counts_properties,
+    scrna_raw_counts_properties,
+)
+from scvi.model.base import UnsupervisedTrainingMixin
+from scvi.module import MULTIVAE
+from scvi.train import AdversarialTrainingPlan, TrainRunner
+from scvi.train._callbacks import SaveBestState
+
+from .base import BaseModelClass, VAEMixin
+from .base._utils import _de_core
+
+Number = Union[int, float]
+logger = logging.getLogger(__name__)
+
+
+class MULTIVI(VAEMixin, UnsupervisedTrainingMixin, BaseModelClass):
+    """
+    Integration of multi-modal and single-modality data. [AshuachGabitto21]_
+
+    MultiVI is used to integrate multiomic datasets with single-modality (expression
+    or accessibility) datasets.
+
+    Parameters
+    ----------
+    adata
+        AnnData object that has been registered via :func:`~scvi.data.setup_anndata`.
+    n_genes
+        The number of gene expression features (genes).
+    n_regions
+        The number of accessibility features (genomic regions).
+    n_hidden
+        Number of nodes per hidden layer. If `None`, defaults to square root
+        of number of regions.
+    n_latent
+        Dimensionality of the latent space. If `None`, defaults to square root
+        of `n_hidden`.
+    n_layers_encoder
+        Number of hidden layers used for encoder NNs.
+    n_layers_decoder
+        Number of hidden layers used for decoder NNs.
+    dropout_rate
+        Dropout rate for neural networks.
+    model_depth
+        Model sequencing depth / library size.
+    region_factors
+        Include region-specific factors in the model.
+    latent_distribution
+        One of
+        * ``'normal'`` - Normal distribution
+        * ``'ln'`` - Logistic normal distribution (Normal(0, I) transformed by softmax)
+    deeply_inject_covariates
+        Whether to deeply inject covariates into all layers of the decoder. If False,
+        covariates will only be included in the input layer.
+    fully_paired
+        allows the simplification of the model if the data is fully paired. Currently ignored.
+    **model_kwargs
+        Keyword args for :class:`~scvi.module.PEAKVAE`
+
+    Notes
+    -----
+    * The model assumes that the features are organized so that all expression features are
+    consecutive, followed by all accessibility features. For example, if the data has 100 genes
+    and 250 genomic regions, the model assumes that the first 100 features are genes, and the
+    next 250 are the regions.
+
+    * The main batch annotation, specified in the `scvi.data.setup_anndata`, should correspond to
+    the modality each cell originated from. This allows the model to focus mixing efforts, using
+    an adversarial component, on mixing the modalities. Other covariates can be specified using
+    the `categorical_covariate_keys` argument.
```

Comment:
```suggestion
    * The model assumes that the features are organized so that all expression features are
       consecutive, followed by all accessibility features. For example, if the data has 100 genes
       and 250 genomic regions, the model assumes that the first 100 features are genes, and the
       next 250 are the regions.

    * The main batch annotation, specified in the `scvi.data.setup_anndata`, should correspond to
       the modality each cell originated from. This allows the model to focus mixing efforts, using
       an adversarial component, on mixing the modalities. Other covariates can be specified using
       the `categorical_covariate_keys` argument.
```